### PR TITLE
feat: Registration flow

### PR DESCRIPTION
**Pull request by atticusofsparta**

### CODE REVIEW

File: src/components/inputs/Counter/Counter.tsx [hunk 1]
```diff
@@ -1,52 +1,71 @@
...
-        date.setFullYear(date.getFullYear() + count);
+        date.setFullYear(date.getFullYear() + leaseDuration);
         break;
       case 'days':
-        date.setDate(date.getDate() + count);
+        date.setDate(date.getDate() + leaseDuration);
         break;
       case 'minutes':
-        date.setHours(date.getHours() + count);
+        date.setHours(date.getHours() + leaseDuration);
```
Comment (dtfiedler) [inline src/components/inputs/Counter/Counter.tsx:68]: in an effort to keep the counter generic, keep `count`. i know you mentioned coming back to it in another PR, but at least at a `TODO` comment so we remember

File: src/components/inputs/Search/SearchBar/SearchBar.tsx [hunk 2]
```diff
@@ -49,15 +59,15 @@ function SearchBar(props: SearchBarProps) {
     // partially reset
     setSearchResult(undefined);
     setShowDefaultText(true);
-    setSubmittedName(undefined);
+    //setSubmittedName(undefined);
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/SearchBar.tsx:62]: nit: remove

File: src/components/inputs/Search/SearchBar/SearchBar.tsx [hunk 3]
```diff
@@ -81,35 +94,50 @@ function SearchBar(props: SearchBarProps) {
     // show updated states based on search result
     const name = searchBarText;
     const searchSuccess = successPredicate(name);
+    console.log(searchSuccess);
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/SearchBar.tsx:97]: nit: remove

File: src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      fee,
+      stage,
+    },
+    dispatchRegistrationState,
+  ] = useRegistrationState();
+  const [priceInfo, setPriceInfo] = useState(false);
+
+  return (
+    <>
+      <div className="register-name-modal center">
```
Comment (dtfiedler) [inline src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx:31]: nit: add a `TODO` to update ant-card and use it here

File: src/components/layout/DeployRegistration/DeployRegistration.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+              title: 'Create an ANT',
+              status: '',
+            },
+            3: {
+              title: 'Register Name',
+              status: '',
+            },
+            4: {
+              title: 'Complete',
+              status: '',
```
Comment (dtfiedler) [inline src/components/layout/DeployRegistration/DeployRegistration.tsx:24]: we're achieving this through the ordering of keys, but it might be good to have `dependsOn: [1, 2]`  and `next` explicitly defined in the object, so we clearly understand the dependencies of each step. not necessary for this PR, but something to keep in mind

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    if (!isArweaveTransactionID(id)) {
+      updateRegisterState({ key: 'setTargetID', value: undefined });
+    }
+  }
+
+  return (
+    <>
+      <div className="register-name-modal center">
+        <div className="section-header">{domain} is available!</div>
+
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:49]: nit: surprised the husky didn't catch these extra spaces 🤔

File: src/components/layout/Workflow/Workflow.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  const [{ stage, isFirstStage, isLastStage }, dispatchRegisterState] =
+    useRegistrationState();
+  const [{ walletAddress, jwk, isSearching }, dispatchGlobalState] =
+    useGlobalState();
+
+  const [nextCondition, setNextCondition] = useState<boolean>(false);
+  const [backCondition, setBackCondition] = useState<boolean>(false);
+
+  useEffect(() => {
+    // eslint-disable-next-line
```
Comment (dtfiedler) [inline src/components/layout/Workflow/Workflow.tsx:18]: did this complain?

File: src/components/pages/Home/Home.tsx [hunk 8]
```diff
@@ -29,29 +36,73 @@ function Home() {
...
+                }
+              />
+            }
+            height={45}
           />
-        }
-      />
+        ),
+        nextCondition: true,
+        backCondition: true,
```
Comment (dtfiedler) [inline src/components/pages/Home/Home.tsx:65]: what is a use case for `backCondition`?

File: src/state/contexts/GlobalState.tsx [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
-import type { ArNSContractState, ArweaveWalletConnector } from '../../types';
+import type { ArNSContractState } from '../../types';
 import type { Action } from '../reducers/GlobalReducer';
 
 export type GlobalState = {
   arnsSourceContract: ArNSContractState;
   gateway: string;
   walletAddress?: string;
-  wallet?: ArweaveWalletConnector;
+  jwk?: JWKInterface;
```
Comment (dtfiedler) [inline src/state/contexts/GlobalState.tsx:11]: don't revert this change, make sure you pull most recent version of develop via `git merge origin/develop`

File: src/state/reducers/RegistrationReducer.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
+import { ArweaveTransactionId } from '../../types';
+import { RegistrationState } from '../contexts/RegistrationState';
+
+export type RegistrationAction =
+  | { type: 'setDomainName'; payload: string }
+  | { type: 'setLeaseDuration'; payload: number }
+  | { type: 'setChosenTier'; payload: number }
```
Comment (dtfiedler) [inline src/state/reducers/RegistrationReducer.ts:7]:
```suggestion
  | { type: 'setTier'; payload: number }
```

File: src/types.ts [hunk 11]
```diff
@@ -22,6 +23,9 @@ export type ArNSDomain = ArNSMapping & ArNSMetaData;
 // TODO: match this to a regex
 export type ArweaveTransactionId = string;
 
+export interface WalletUploadSource {
+  getWallet(e: any): JWKInterface | void;
+}
```
Comment (dtfiedler) [inline src/types.ts:28]: pull in most recent develop

File: src/utils/searchUtils.ts [hunk 12]
```diff
@@ -20,6 +20,7 @@ export function isArNSDomainNameAvailable({
   if (!name || Object.keys(records).includes(name)) {
     return false;
   }
+  console.log(records);
```
Comment (dtfiedler) [inline src/utils/searchUtils.ts:23]: nit: remove

File: src/utils/searchUtils.ts [hunk 13]
```diff
@@ -59,3 +60,15 @@ export function calculateArNSNamePrice({
...
+export function isArweaveTransactionID(id: string) {
+  if (!id) {
+    return false;
+  }
+  if (!ARNS_TXID_REGEX.test(id)) {
+    return false;
+  }
+  if (ARNS_TXID_REGEX.test(id)) {
+    return true;
+  }
```
Comment (dtfiedler) [inline src/utils/searchUtils.ts:73]:
```suggestion
  if (!id) {
    return false;
  }
  if (!ARNS_TXID_REGEX.test(id)) {
    return false;
  }
   
  return true;
```

File: src/components/inputs/Counter/Counter.tsx (hunk 1 above)
Comment (atticusofsparta) [reply src/components/inputs/Counter/Counter.tsx:68]: added todo

File: src/components/inputs/Search/SearchBar/SearchBar.tsx (hunk 2 above)
Comment (atticusofsparta) [reply src/components/inputs/Search/SearchBar/SearchBar.tsx:62]: removed

File: src/components/inputs/Search/SearchBar/SearchBar.tsx (hunk 3 above)
Comment (atticusofsparta) [reply src/components/inputs/Search/SearchBar/SearchBar.tsx:97]: removed

File: src/components/inputs/progress/RegistrationProgress/RegistrationProgress.tsx [hunk 14]
```diff
@@ -0,0 +1,155 @@
+import { useIsMobile } from '../../../../hooks';
+import {
+  AlertOctagonIcon,
+  AlertTriangleIcon,
+  CheckmarkIcon,
+} from '../../../icons';
+import './styles.css';
+
+function RegistrationProgress({
```
[the remark this answers is not in the document]
Comment (atticusofsparta) [reply src/components/inputs/progress/RegistrationProgress/RegistrationProgress.tsx:9]: renamed to StagesProgressBar

File: src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx (hunk 4 above)
Comment (atticusofsparta) [reply src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx:31]: todo added

File: src/components/layout/DeployRegistration/DeployRegistration.tsx (hunk 5 above)
Comment (atticusofsparta) [reply src/components/layout/DeployRegistration/DeployRegistration.tsx:24]: added todo

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx (hunk 6 above)
Comment (atticusofsparta) [reply src/components/layout/RegisterNameForm/RegisterNameForm.tsx:49]: human formatter activated, spaces destroyed

File: src/components/layout/Workflow/Workflow.tsx (hunk 7 above)
Comment (atticusofsparta) [reply src/components/layout/Workflow/Workflow.tsx:18]: yes, no-unused-vars on the key from object.entries

File: src/components/inputs/progress/RegistrationProgress/RegistrationProgress.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (atticusofsparta) [reply src/components/inputs/progress/RegistrationProgress/RegistrationProgress.tsx:9]: scratch that, StepProgressBar

File: src/components/pages/Home/Home.tsx (hunk 8 above)
Comment (atticusofsparta) [reply src/components/pages/Home/Home.tsx:65]: For when you dont want the user to be able to go back, or when you have an alternative display component or conditional that handles going back a step. For example, during deployment have a transaction, it could be found that a user suddenly has insufficient funds for gas or payment, the current comp can then handle the back condition. Currently this bool simply handles the actual rendering of the back button instead of the condition of the back button being available. A better solution would be to come back in another pr and advance this to a predicate for both the nextCondition and backCondition - however with the current pattern that is difficult to lay out because of the components being initialized outside of the workflow component - maybe we can use some sort of generic even emitter on the workflow buttons where they know "ok component "search" emitted its onSubmit state, ()=> predicate based on constants from nextCondition... like arg0 === false/string with x chars or something.

File: src/state/contexts/GlobalState.tsx (hunk 9 above)
Comment (atticusofsparta) [reply src/state/contexts/GlobalState.tsx:11]: wierd.... merged... file not updating... must have merged then overwrote it....

Comment (atticusofsparta) [reply src/state/contexts/GlobalState.tsx:11]: anyway, added it

File: src/state/reducers/RegistrationReducer.ts (hunk 10 above)
Comment (atticusofsparta) [reply src/state/reducers/RegistrationReducer.ts:7]: updated

File: src/types.ts (hunk 11 above)
Comment (atticusofsparta) [reply src/types.ts:28]: fixed

File: src/utils/searchUtils.ts (hunk 12 above)
Comment (atticusofsparta) [reply src/utils/searchUtils.ts:23]: removed

File: src/utils/searchUtils.ts (hunk 13 above)
Comment (atticusofsparta) [reply src/utils/searchUtils.ts:73]: changed

File: src/state/contexts/GlobalState.tsx (hunk 9 above)
Comment (dtfiedler) [reply src/state/contexts/GlobalState.tsx:12]: we don't want jwk at all, we'll use the `ArweaveWalletConnector` or something different here

File: src/state/contexts/GlobalState.tsx [hunk 15]
```diff
@@ -17,6 +19,7 @@ const initialState: GlobalState = {
   arnsSourceContract: { records: {}, fees: {} },
   gateway: 'arweave.dev',
   walletAddress: undefined,
+  jwk: undefined,
```
Comment (dtfiedler) [inline src/state/contexts/GlobalState.tsx:22]: remove

File: src/components/cards/TierCard/TierCard.tsx [hunk 16]
```diff
@@ -1,14 +1,16 @@
+import { useRegistrationState } from '../../../state/contexts/RegistrationState';
 import { TierCardProps } from '../../../types';
 import { TIER_DATA } from '../../../utils/constants';
 import { CircleCheck } from '../../icons';
 import './styles.css';
 
-function TierCard({ tier, setTier, selectedTier }: TierCardProps) {
+function TierCard({ currentTier }: TierCardProps) {
```
Comment (dtfiedler) [inline src/components/cards/TierCard/TierCard.tsx:7]: nit: `selectedTier`

File: src/components/inputs/Search/SearchBar/styles.css [hunk 17]
```diff
@@ -42,19 +47,20 @@
   color: var(--placeholder-text);
   text-transform: none;
 }
-.search-button {
+.searchbutton {
```
Comment (dtfiedler) [inline src/components/inputs/Search/SearchBar/styles.css:50]: we're still using `search-button` in the class name where this is used

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+            <Dropdown
+              showSelected={true}
+              showChevron={true}
+              selected={antFlow}
+              setSelected={setAntFlow}
+              options={{
+                'Create an Arweave Name Token (ANT) for me':
+                  'Create an Arweave Name Token (ANT) for me',
+                'I have an Arweave Name Token (ANT) I want to use':
+                  'I have an Arweave Name Token (ANT) I want to use',
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:61]:
```suggestion
                'create-ant':
                  'Create an Arweave Name Token (ANT) for me',
                'use-existing':
                  'I have an Arweave Name Token (ANT) I want to use',
```

File: src/components/layout/Workflow/Workflow.tsx (hunk 7 above)
Comment (dtfiedler) [reply src/components/layout/Workflow/Workflow.tsx:18]: FWIW I think you can set it an empty array to avoid this

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+              <></>
+            )}
+          </div>
+          {/**add a tag component for multiple controllers seperated by "," commas. In the future might have an address book. Possibly suggest other accounts from arconnect. */}
+          <input
+            className="data-input center"
+            type="text"
+            placeholder="Controllers"
+            onChange={(e) => handleTargetID(e.target.value)}
+            maxLength={43}
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:120]: disallow whitespace and special chars (except `-` and `_`) for all of these inputs

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+            />
+            <input
+              className="data-input center"
+              type="text"
+              placeholder="Ticker*"
+            />
+            <Dropdown
+              showSelected={true}
+              showChevron={true}
+              selected={ttl}
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:136]:
```suggestion
              selected=`{ttl} secs`
```

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+              }
+              options={{
+                '100s': 100,
+                '200s': 200,
+                '300s': 300,
+                '400s': 400,
+                '500s': 500,
+                '600s': 600,
+                '700s': 700,
+                '800s': 800,
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:148]: change `s` to `secs` - i know this probably conflicts with design but it's not clear what this is setting

File: src/components/pages/Home/Home.tsx (hunk 8 above)
Comment (dtfiedler) [reply src/components/pages/Home/Home.tsx:65]: i understand the intent, but still circling my head around an ideal implementation. fine to leave for now and address in another PR

File: src/components/cards/TierCard/TierCard.tsx (hunk 16 above)
Comment (atticusofsparta) [reply src/components/cards/TierCard/TierCard.tsx:7]: Its not really selected - the selected tier is "tier" from the useRegistrationState. currentTier is more like.... this.tierData. maybe tierNumber / tierName would be more apt?

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx (hunk 18 above)
Comment (atticusofsparta) [reply src/components/layout/RegisterNameForm/RegisterNameForm.tsx:61]: added

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx (hunk 19 above)
Comment (atticusofsparta) [reply src/components/layout/RegisterNameForm/RegisterNameForm.tsx:120]: fixed

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx (hunk 21 above)
Comment (atticusofsparta) [reply src/components/layout/RegisterNameForm/RegisterNameForm.tsx:148]: changed, but this *should* be getting changed to a counter comp in a future pr

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx (hunk 20 above)
Comment (atticusofsparta) [reply src/components/layout/RegisterNameForm/RegisterNameForm.tsx:136]: added

File: src/components/layout/RegisterNameForm/RegisterNameForm.tsx [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+                value={targetID}
+                onChange={(e) => handleTargetID(e)}
+                maxLength={43}
+                style={
+                  targetID && isArweaveTransactionID(targetID)
+                    ? { border: '2px solid var(--success-green)' }
+                    : targetID && !isArweaveTransactionID(targetID)
+                    ? { border: '2px solid var(--error-red)' }
+                    : {}
+                }
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/RegisterNameForm.tsx:59]: image


also, the tx id for a current ANT fails the validation `gh673M0Koh941OIITVXl9hKabRaYWABQUedZxW-swIA`

File: src/components/layout/RegisterNameForm/styles.css [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  max-width: 1000px;
+  min-width: 320px;
+  width: 80%;
+  height: fit-content;
+  margin: auto;
+  padding: 2em 2em;
+  gap: 2em;
+  background: var(--card-bg);
+  border: 1px solid #323232;
+  border-radius: 32px;
```
Comment (dtfiedler) [inline src/components/layout/RegisterNameForm/styles.css:13]: i personally don't care for the shaded box around the workflow, but know that's what figma portrays - let's see what the feedback is at the next demo for this

File: src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+            <b>{controllers.length ? controllers : 'not set'}</b>
+          </span>
+          <span className="detail">
+            Owner:&nbsp;<b>{walletAddress ? walletAddress : 'not set'}</b>
+          </span>
+          <span className="detail">
+            ttlSeconds:&nbsp;<b>{ttl ? ttl : 'not set'}</b>
+          </span>
+          <span className="detail">
+            Subdomains:&nbsp;<b>Up to 100</b>
```
Comment (dtfiedler) [inline src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx:62]: shouldn't this depend on the tier selection?

File: src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+          </span>
+          <span className="detail">
+            Nickname*:&nbsp;<b>{nickname ? nickname : 'not set'}</b>
+          </span>
+          <span className="detail">
+            Ticker*:&nbsp;<b>{ticker ? ticker : 'not set'}</b>
+          </span>
+          <span className="detail">
+            Controllers*:&nbsp;
+            <b>{controllers.length ? controllers : 'not set'}</b>
```
Comment (dtfiedler) [inline src/components/layout/ConfirmRegistration/ConfirmRegistration.tsx:53]: this are not getting set:

```
domain: wallets
taretId: gh673M0Koh941OIITVXl9hKabRaYWABQUedZxW-swIA
```

File: src/components/layout/Workflow/Workflow.tsx [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+      payload: true,
+    });
+  }
+
+  return (
+    <>
+      {/* eslint-disable-next-line */}
+      {Object.entries(stages).map(([key, value], index) => {
+        if (index === stage) {
+          return value.component;
```
Comment (dtfiedler) [inline src/components/layout/Workflow/Workflow.tsx:40]: image
```suggestion
          return React.clone(value.component, {
              key
          })
```

File: src/components/layout/Workflow/Workflow.tsx [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    dispatchGlobalState({
+      type: 'setShowConnectWallet',
+      payload: true,
+    });
+  }
+
+  return (
+    <>
+      {/* eslint-disable-next-line */}
+      {Object.entries(stages).map(([key, value], index) => {
```
Comment (dtfiedler) [inline src/components/layout/Workflow/Workflow.tsx:38]:
```suggestion
      {Object.entries(stages).map(([key], index) => {
```

Comment (dtfiedler) [inline src/components/layout/Workflow/Workflow.tsx:38]:
```suggestion
      {Object.entries(stages).map(([key], index) => {
```

File: src/components/layout/Workflow/Workflow.tsx [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+    useRegistrationState();
+  const [{ walletAddress, isSearching }, dispatchGlobalState] =
+    useGlobalState();
+
+  const [nextCondition, setNextCondition] = useState<boolean>(false);
+  const [backCondition, setBackCondition] = useState<boolean>(false);
+
+  useEffect(() => {
+    // eslint-disable-next-line
+    Object.entries(stages).map(([key, value], index) => {
```
Comment (dtfiedler) [inline src/components/layout/Workflow/Workflow.tsx:19]:
```suggestion
    Object.entries(stages).map(([], index) => {
```
